File: lib/features/taxes/export/build_1040_summary.dart
/// Pure structuring of a TaxCalculationResult into print/PDF-ready groups. No
/// Flutter, no I/O — the printable view and any future PDF generator both read
/// this.
library;

import 'package:finnacalc/features/taxes/engine/types/filing.dart';
import 'package:finnacalc/features/taxes/engine/types/result.dart';

class SummaryLine {
  SummaryLine({required this.label, required this.amount, this.formRef});

  String label;
  double amount;
  String? formRef;
}

class SummaryGroup {
  SummaryGroup({required this.title, required this.lines});

  String title;
  List<SummaryLine> lines;
}

class Form1040SummaryHeadline {
  Form1040SummaryHeadline({
    required this.label,
    required this.amount,
    required this.owes,
  });

  String label;
  double amount;
  bool owes;
}

class Form1040SummaryState {
  Form1040SummaryState({
    required this.name,
    required this.hasIncomeTax,
    required this.tax,
    required this.refundOrOwed,
    this.note,
  });

  String name;
  bool hasIncomeTax;
  double tax;
  double refundOrOwed;
  String? note;
}

class Form1040Summary {
  Form1040Summary({
    required this.filingStatusLabel,
    required this.headline,
    required this.groups,
    this.state,
  });

  final int taxYear = 2024;
  String filingStatusLabel;
  Form1040SummaryHeadline headline;
  List<SummaryGroup> groups;
  Form1040SummaryState? state;
}

const Map<FilingStatus, String> _filingLabels = {
  FilingStatus.single: 'Single',
  FilingStatus.mfj: 'Married filing jointly',
  FilingStatus.mfs: 'Married filing separately',
  FilingStatus.hoh: 'Head of household',
  FilingStatus.qss: 'Qualifying surviving spouse',
};

/// Title-case a camelCase credit key, e.g. "childTaxCredit" → "Child tax
/// credit".
String labelizeCredit(String key) {
  final spaced = key
      .replaceAllMapped(RegExp(r'([A-Z])'), (m) => ' ${m[1]}')
      .toLowerCase();
  return spaced.substring(0, 1).toUpperCase() + spaced.substring(1);
}

Form1040Summary build1040Summary(TaxCalculationResult r) {
  final List<SummaryGroup> groups = [];

  groups.add(
    SummaryGroup(
      title: 'Income',
      lines: [
        SummaryLine(
          label: 'Total income',
          amount: r.totalIncome,
          formRef: '1040 line 9',
        ),
        SummaryLine(
          label: 'Adjustments to income',
          amount: r.totalAdjustments,
          formRef: 'Schedule 1',
        ),
        SummaryLine(
          label: 'Adjusted gross income (AGI)',
          amount: r.agi,
          formRef: '1040 line 11',
        ),
      ],
    ),
  );

  final List<SummaryLine> deductionLines = [
    SummaryLine(
      label: r.deductionUsed == 'itemized'
          ? 'Itemized deductions'
          : 'Standard deduction',
      amount: r.deductionAmount,
      formRef: '1040 line 12',
    ),
  ];
  if (r.qbiDeduction > 0) {
    deductionLines.add(
      SummaryLine(
        label: 'QBI deduction',
        amount: r.qbiDeduction,
        formRef: '1040 line 13',
      ),
    );
  }
  deductionLines.add(
    SummaryLine(
      label: 'Taxable income',
      amount: r.taxableIncome,
      formRef: '1040 line 15',
    ),
  );
  groups.add(SummaryGroup(title: 'Deductions', lines: deductionLines));

  final List<SummaryLine> taxLines = [
    SummaryLine(
      label: 'Income tax',
      amount: r.regularTax,
      formRef: '1040 line 16',
    ),
  ];
  if (r.amt > 0) {
    taxLines.add(
      SummaryLine(
        label: 'Alternative minimum tax',
        amount: r.amt,
        formRef: 'Schedule 2',
      ),
    );
  }
  for (final entry in r.nonrefundableCredits.entries) {
    taxLines.add(
      SummaryLine(
        label: '− ${labelizeCredit(entry.key)}',
        amount: -entry.value,
      ),
    );
  }
  if (r.seTax > 0) {
    taxLines.add(
      SummaryLine(
        label: 'Self-employment tax',
        amount: r.seTax,
        formRef: 'Schedule 2',
      ),
    );
  }
  if (r.additionalMedicareTax > 0) {
    taxLines.add(
      SummaryLine(
        label: 'Additional Medicare tax',
        amount: r.additionalMedicareTax,
      ),
    );
  }
  if (r.niit > 0) {
    taxLines.add(
      SummaryLine(label: 'Net investment income tax', amount: r.niit),
    );
  }
  taxLines.add(
    SummaryLine(
      label: 'Total tax',
      amount: r.totalTax,
      formRef: '1040 line 24',
    ),
  );
  groups.add(SummaryGroup(title: 'Tax & credits', lines: taxLines));

  final List<SummaryLine> payLines = [];
  for (final entry in r.refundableCredits.entries) {
    payLines.add(
      SummaryLine(label: labelizeCredit(entry.key), amount: entry.value),
    );
  }
  payLines.add(
    SummaryLine(
      label: 'Total payments & refundable credits',
      amount: r.totalPayments,
      formRef: '1040 line 33',
    ),
  );
  groups.add(SummaryGroup(title: 'Payments', lines: payLines));

  return Form1040Summary(
    filingStatusLabel: _filingLabels[r.filingStatus]!,
    headline: Form1040SummaryHeadline(
      label: r.owes ? 'Estimated balance due' : 'Estimated federal refund',
      amount: r.refundOrOwed.abs(),
      owes: r.owes,
    ),
    groups: groups,
    state: r.state != null && r.state!.supported
        ? Form1040SummaryState(
            name: r.state!.name,
            hasIncomeTax: r.state!.hasIncomeTax,
            tax: r.state!.tax,
            refundOrOwed: r.state!.refundOrOwed,
            note: r.state!.note,
          )
        : null,
  );
}

File: app/src/main/kotlin/com/finnacalc/android/features/taxes/engine/Build1040Summary.kt
/**
 * Build1040Summary.kt
 *
 * Port of iOS Features/Taxes/Engine/Build1040Summary.swift.
 *
 * Pure structuring of a TaxCalculationResult into print/share-ready groups. No
 * Compose, no I/O — the printable view and any future PDF generator both read
 * this. Same group/line ordering, labels, form references, and conditional
 * inclusion rules as the Swift and TypeScript originals.
 *
 * Credit entries are emitted in sorted-key order, as in the Swift port, so the
 * summary is deterministic regardless of map iteration order.
 */

package com.finnacalc.android.features.taxes.engine

import kotlin.math.abs

data class SummaryLine(val label: String, val amount: Double, val formRef: String? = null)

data class SummaryGroup(val title: String, val lines: List<SummaryLine>)

data class Form1040Headline(val label: String, val amount: Double, val owes: Boolean)

data class Form1040SummaryState(
    val name: String,
    val hasIncomeTax: Boolean,
    val tax: Double,
    val refundOrOwed: Double,
    val note: String? = null,
)

data class Form1040Summary(
    val taxYear: Int = 2025,
    val filingStatusLabel: String,
    val headline: Form1040Headline,
    val groups: List<SummaryGroup>,
    val state: Form1040SummaryState? = null,
)

private val filingLabels: Map<FilingStatus, String> = mapOf(
    FilingStatus.Single to "Single",
    FilingStatus.Mfj to "Married filing jointly",
    FilingStatus.Mfs to "Married filing separately",
    FilingStatus.Hoh to "Head of household",
    FilingStatus.Qss to "Qualifying surviving spouse",
)

/** Title-case a camelCase credit key: "childTaxCredit" → "Child tax credit". */
private fun labelizeCredit(key: String): String {
    val spaced = buildString {
        for (ch in key) {
            if (ch.isUpperCase()) append(' ')
            append(ch)
        }
    }.lowercase()
    if (spaced.isEmpty()) return spaced
    return spaced.first().uppercase() + spaced.drop(1)
}

fun build1040Summary(r: TaxCalculationResult): Form1040Summary {
    val groups = mutableListOf<SummaryGroup>()

    groups.add(
        SummaryGroup(
            "Income",
            listOf(
                SummaryLine("Total income", r.totalIncome, "1040 line 9"),
                SummaryLine("Adjustments to income", r.totalAdjustments, "Schedule 1"),
                SummaryLine("Adjusted gross income (AGI)", r.agi, "1040 line 11"),
            )
        )
    )

    val deductionLines = mutableListOf(
        SummaryLine(
            if (r.deductionUsed == DeductionUsed.Itemized) "Itemized deductions" else "Standard deduction",
            r.deductionAmount,
            "1040 line 12",
        )
    )
    if (r.qbiDeduction > 0) {
        deductionLines.add(SummaryLine("QBI deduction", r.qbiDeduction, "1040 line 13"))
    }
    deductionLines.add(SummaryLine("Taxable income", r.taxableIncome, "1040 line 15"))
    groups.add(SummaryGroup("Deductions", deductionLines))

    val taxLines = mutableListOf(SummaryLine("Income tax", r.regularTax, "1040 line 16"))
    if (r.amt > 0) {
        taxLines.add(SummaryLine("Alternative minimum tax", r.amt, "Schedule 2"))
    }
    for (key in r.nonrefundableCredits.keys.sorted()) {
        val amount = r.nonrefundableCredits.getValue(key)
        taxLines.add(SummaryLine("− ${labelizeCredit(key)}", -amount))
    }
    if (r.seTax > 0) {
        taxLines.add(SummaryLine("Self-employment tax", r.seTax, "Schedule 2"))
    }
    if (r.additionalMedicareTax > 0) {
        taxLines.add(SummaryLine("Additional Medicare tax", r.additionalMedicareTax))
    }
    if (r.niit > 0) {
        taxLines.add(SummaryLine("Net investment income tax", r.niit))
    }
    taxLines.add(SummaryLine("Total tax", r.totalTax, "1040 line 24"))
    groups.add(SummaryGroup("Tax & credits", taxLines))

    val payLines = mutableListOf<SummaryLine>()
    for (key in r.refundableCredits.keys.sorted()) {
        payLines.add(SummaryLine(labelizeCredit(key), r.refundableCredits.getValue(key)))
    }
    payLines.add(SummaryLine("Total payments & refundable credits", r.totalPayments, "1040 line 33"))
    groups.add(SummaryGroup("Payments", payLines))

    val s = r.state
    val stateBlock = if (s != null && s.supported) {
        Form1040SummaryState(
            name = s.name,
            hasIncomeTax = s.hasIncomeTax,
            tax = s.tax,
            refundOrOwed = s.refundOrOwed,
            note = s.note,
        )
    } else null

    return Form1040Summary(
        taxYear = 2025,
        filingStatusLabel = filingLabels[r.filingStatus] ?: "",
        headline = Form1040Headline(
            label = if (r.owes) "Estimated balance due" else "Estimated federal refund",
            amount = abs(r.refundOrOwed),
            owes = r.owes,
        ),
        groups = groups,
        state = stateBlock,
    )
}
